datafusion/sql/src/expr/value.rs

```diff
@@ -151,16 +151,51 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
         let data_types: HashSet<DataType> =
```

Comment:
Yeah, I agree the use of a `HashSet` here is going to be tough because it will lose the argument order (and thus you can't pick the first non null type deterministically 

What about keeping track of the types in both the map and a Vec, like

```
let data_types: Vec<DataType>= vec![];
let seen_types: HashSet<DataType> = HashSet::new();
```

Then I think you can use `data_types` below as you'll know it is non duplicated